File: InterviewBit/DP/Largest Sqaure and Rectangles/optimized_sqaure1.py
#!/anaconda3/bin/python3
'''
Runtime: 72 ms, faster than 97.69% of Python3 online submissions for Maximal Square.
Memory Usage: 13.9 MB, less than 68.14% of Python3 online submissions for Maximal Square.
'''
import math
class Solution:
    def maximalSquare(self,A):
        if len(A)==0 or len(A[0])==0:
            return 0
        pre=[0]*len(A[0])
        current=[0]*len(A[0])
        max=-math.inf
        for i in range(len(A)):
            for j in range(len(A[i])):
                if i==0 or j==0:
                    current[j]=int(A[i][j])
                elif int(A[i][j])==1:
                    current[j]=1+min((pre[j]),(pre[j-1]),(current[j-1]))
                if max<current[j]:
                    max=current[j]
            pre=current
            current=[0]*len(A[0])
        return max*max


sol_obj=Solution()
A=[[1,1,1,0],[1,1,1,1],[1,1,0,0]]
A=[["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
#A=[["1"]]
#A=[["1","1","1","1"],["1","1","1","1"],["1","1","1","1"]]
out=sol_obj.maximalSquare(A)
print (out)

File: InterviewBit/DP/Largest Sqaure and Rectangles/optimized_sqaure1.cpp
#include <vector>
#include <iostream>
using namespace std;
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if (matrix.empty()) {
            return 0;
        }
        int m = matrix.size(), n = matrix[0].size(), sz = 0;
        vector<int> pre(n, 0), cur(n, 0);
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!i || !j ) {
                    cur[j] = matrix[i][j] - '0';
                } else if (matrix[i][j] == '1'){
                    cur[j] = min(pre[j - 1], min(pre[j], cur[j - 1])) + 1;
                }
                sz = max(cur[j], sz);
            }
            fill(pre.begin(), pre.end(), 0);
            swap(pre, cur);
        }
        return sz * sz;
    }
};
int main(){
  Solution sol_obj;
  vector<vector<char> > vect{{ '1','0','1','0','0'},
  { '1','0','1','1','1'},
  { '1','1','1','1','1'},
  { '1','0','0','1','0'}
};
int out=sol_obj.maximalSquare(vect);
cout<<out<<endl;
//int a='1'-'0'; // hence chars can be operated and assigned to ints
//cout << a<<endl;
//int b='0'-'0';
//cout << b<<endl;
return 0;
}
